### ISSUE TITLE
Storage Provider - Failed to start Orleans silo - Please Help

### ISSUE THREAD

**Issue by combrinckd** [dotnet/orleans#1016]
I've been trying to figure this out for the whole day now, looking at various posts like https://github.com/dotnet/orleans/issues/786 and then the default tutorials and documentations etc.
However after all this I still cant figure out why I can't start the silo with the storage provider.
I have made sure that the corresponding assembly is in the execution path etc as per posts and documentation etc.
I am not sure what to look for any more. Please see log errors, code etc below.

**The Error and log abstraction**
Exc level 0: System.AggregateException: One or more errors occurred.
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(TimeSpan timeout)
   at Orleans.OrleansTaskExtentions.WaitWithThrow(Task task, TimeSpan timeout)
   at Orleans.Runtime.Silo.DoStart()
   at Orleans.Runtime.Silo.Start()
   at Orleans.Runtime.Host.SiloHost.StartOrleansSilo()
Exc level 1: Orleans.Runtime.OrleansException: Provider of type CCMS.StorageProviders.SqlServerStorageProvider name SqlStorage was not loaded.
   at Orleans.Providers.ProviderLoader`1.ValidateProviders()
   at Orleans.Providers.ProviderLoader`1.LoadProviders(IDictionary`2 configs, IProviderManager providerManager)
   at Orleans.Runtime.Storage.StorageProviderManager.LoadStorageProviders(IDictionary`2 configs)
   at Orleans.Runtime.Silo.<DoStart>b__11()
   at Orleans.Runtime.Scheduler.SchedulerExtensions.<>c__DisplayClassa.<<QueueTask>b__8>d__c.MoveNext()

**AND**

[2015-11-11 12:20:46.171 GMT    10	ERROR  	101716	AssemblyLoader.Silo	127.0.0.1:11111]	!!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.	
Exc level 0: System.AggregateException: A ReflectionTypeLoadException has been thrown. The original exception and the contents of the LoaderExceptions property have been aggregated for your convenence.
   at Orleans.Runtime.Utils.Flatten(ReflectionTypeLoadException rtle)
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass1.<NewCriterion>b__0(Assembly assembly, IEnumerable`1& assemblyComplaints)
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass5.<.ctor>b__3(Object input, IEnumerable`1& complaints)
   at Orleans.Runtime.AssemblyLoaderCriterion.EvaluateCandidate(Object input, IEnumerable`1& complaints)
   at Orleans.Runtime.AssemblyLoader.ShouldLoadAssembly(String pathName)
Exc level 1: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.Assembly.GetTypes()
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass1.<NewCriterion>b__0(Assembly assembly, IEnumerable`1& assemblyComplaints)
Exc level 2: System.TypeLoadException: Method 'ReadStateAsync' in type 'CCMS.StorageProviders.SqlServerStorageProvider' from assembly 'CCMS.StorageProviders, Version=1.0.0.0, Culture=neutral, PublicKeyToken=8ace0fb1850942f7' does not have an implementation.
Exc level 1: System.TypeLoadException: Method 'ReadStateAsync' in type 'CCMS.StorageProviders.SqlServerStorageProvider' from assembly 'CCMS.StorageProviders, Version=1.0.0.0, Culture=neutral, PublicKeyToken=8ace0fb1850942f7' does not have an implementation.
[2015-11-11 12:20:46.579 GMT    10	ERROR  	101716	AssemblyLoader.Silo	127.0.0.1:11111]	!!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.	
Exc level 0: System.AggregateException: A ReflectionTypeLoadException has been thrown. The original exception and the contents of the LoaderExceptions property have been aggregated for your convenence.
   at Orleans.Runtime.Utils.Flatten(ReflectionTypeLoadException rtle)
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass1.<NewCriterion>b__0(Assembly assembly, IEnumerable`1& assemblyComplaints)
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass5.<.ctor>b__3(Object input, IEnumerable`1& complaints)
   at Orleans.Runtime.AssemblyLoaderCriterion.EvaluateCandidate(Object input, IEnumerable`1& complaints)
   at Orleans.Runtime.AssemblyLoader.ShouldLoadAssembly(String pathName)
Exc level 1: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.Assembly.GetTypes()
   at Orleans.Runtime.AssemblyLoaderReflectionCriterion.<>c__DisplayClass1.<NewCriterion>b__0(Assembly assembly, IEnumerable`1& assemblyComplaints)
Exc level 2: System.TypeLoadException: Method 'ReadStateAsync' in type 'CCMS.StorageProviders.SqlServerStorageProvider' from assembly 'CCMS.StorageProviders, Version=1.0.0.0, Culture=neutral, PublicKeyToken=8ace0fb1850942f7' does not have an implementation.
Exc level 1: System.TypeLoadException: Method 'ReadStateAsync' in type 'CCMS.StorageProviders.SqlServerStorageProvider' from assembly 'CCMS.StorageProviders, Version=1.0.0.0, Culture=neutral, PublicKeyToken=8ace0fb1850942f7' does not have an implementation.

**Here is my provider**
With required implementations etc.!!

    public class SqlServerStorageProvider : IStorageProvider
    {
        public Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
        {
            Name = name;
            if (string.IsNullOrWhiteSpace(config.Properties["DataConnectionString"]))
                throw new ArgumentException("DataConnectionString property not set");
            _connectionString = config.Properties["DataConnectionString"];

            return TaskDone.Done;
        }

        public string Name { get; private set; }

        public Task Close()
        {
            return TaskDone.Done;
        }

        public Task ReadStateAsync(string grainType, GrainReference grainReference, GrainState grainState)
        {
            var key = GetKey(grainType, grainReference);
            using (SqlConnection con = new SqlConnection(_connectionString))
            using (SqlCommand cmd = new SqlCommand("select data from GrainState where [key] = @key", con))
            {
                con.Open();
                cmd.Parameters.AddWithValue("key", key);
                object data = cmd.ExecuteScalar();
                if (data != null && data != DBNull.Value)
                {
                    var grain = JavaScriptHelper.Deserialize<GrainState>((string)data, grainState.GetType());
                    grainState.SetAll(grain.AsDictionary());
                }
            }

            return TaskDone.Done;
        }

        public Task WriteStateAsync(string grainType, GrainReference grainReference, GrainState grainState)
        {
            var data = JavaScriptHelper.Serialize(grainState.AsDictionary());
            var key = GetKey(grainType, grainReference);
            using (SqlConnection con = new SqlConnection(_connectionString))
            using (SqlCommand cmd = new SqlCommand("update GrainState set data = @data where [key] = @key", con))
            {
                con.Open();
                cmd.Parameters.AddWithValue("key", key);
                cmd.Parameters.AddWithValue("data", data);
                if (cmd.ExecuteNonQuery() == 0)
                {
                    cmd.CommandText = "insert into GrainState([key], data) values(@key, @data)";
                    cmd.ExecuteNonQuery();
                }
            }
            return TaskDone.Done;
        }

        private static string GetKey(string grainType, GrainReference grainReference)
        {
            string key;
            grainReference.GetPrimaryKey(out key);
            return grainType + ":" + key;
        }

        public Task ClearStateAsync(string grainType, GrainReference grainReference, GrainState grainState)
        {
            var key = GetKey(grainType, grainReference);
            using (SqlConnection con = new SqlConnection(_connectionString))
            using (SqlCommand cmd = new SqlCommand("delete from GrainState where [key] = @key", con))
            {
                con.Open();
                cmd.Parameters.AddWithValue("key", key);
                cmd.ExecuteNonQuery();
            }
            return TaskDone.Done;
        }

        public Logger Log { get; set; }

        private string _connectionString;
    }

**The Config entry**

    <?xml version="1.0" encoding="utf-8"?>
    <OrleansConfiguration xmlns="urn:orleans">
      <Globals>
        <StorageProviders>
          <Provider Type="Orleans.Storage.MemoryStorage" Name="MemoryStore" />
          <Provider Type="CCMS.StorageProviders.SqlServerStorageProvider" Name="SqlStorage" DataConnectionString="Data Source=........" />
        </StorageProviders>
        <SeedNode Address="localhost" Port="11111" />
    	<Application>
    		<Defaults>
    			<Deactivation AgeLimit="24hr"/>
    		</Defaults>
    	</Application>
      </Globals>
      <Defaults>
        <Networking Address="localhost" Port="11111" />
        <ProxyingGateway Address="localhost" Port="30000" />
        <Tracing DefaultTraceLevel="Info" TraceToConsole="true" TraceToFile="{0}-{1}.log">
          <TraceLevelOverride LogPrefix="Application" TraceLevel="Info" />
        </Tracing>
        <Statistics MetricsTableWriteInterval="30s" PerfCounterWriteInterval="30s"    LogWriteInterval="300s" WriteLogStatisticsToTable="true"/>
      </Defaults>
      <Override Node="Primary">
        <Networking Address="localhost" Port="11111" />
        <ProxyingGateway Address="localhost" Port="30000" />
      </Override>
    </OrleansConfiguration>

**Comment by sergeybykov** [comment]
Is it possible that the provider is compiled against a different version of Orleans or that a wrong version of 'CCMS.StorageProviders.dll gets deployed with the silo?

**Comment by combrinckd** [comment]
It is, but I have checked this multiple times as well. Even ensuring the correct version of the dlls are in place by manually building and copying the "provider project" to the silo startup folder before running.
All my projects are on Orleans v1.0.10

**Comment by combrinckd** [comment]
Ok, so revisiting the issue after leaving it for a day, I decided to check all references one by one again. It turns out that for some arb reason that the orleansproviders dll was still v1.0.9 on the host project even though I upgraded to v1.0.10 through nuget a couple of weeks back and the packages.config had the correct version listed. Everything ran fine like this until I added the custom SQL provider though. (Not sure how I missed that one.. :)
So I removed all references, uninstalled nuget packages and installed them again, now it runs like a dream....
At least someone else could use the provider code listed above if they want. :)
